### DISCUSSION TITLE
Disparity in prices between uniswap and wrapper

### DISCUSSION THREAD

**Discussion by Marsh-James** [uniswap-python/uniswap-python#182] [Q&A]
Been testing a few coins to get a sense of the accuracy of the wrapper. I've noticed a difference across a few coins and am curious for others to give it a test. On average I'm seeing a 0%<x<0.5% delta.

Here's my code, please do comment if I've implemented it poorly or give it a test.

```
# Addresses sourced from: https://info.uniswap.org/#/tokens
ETH = '0xC02aaA39b223FE8D0A0e5C4F27eAD9083C756Cc2'
USDT = '0xdAC17F958D2ee523a2206206994597C13D831ec7'
DAI = '0x6B175474E89094C44Da98b954EedeAC495271d0F'

# CONFIGURE UNISWAP
# uniswap_keys = get_keys('../bin/uniswap.secret')
uniswap_exchange = uniswap.Uniswap(
    address='', #uniswap_keys['key'],
    private_key='', #uniswap_keys['secret'],
    provider='', #'https://mainnet.infura.io/v3/<PROJECT>',
    version=3,
    default_slippage=0.01
)

DAI_SWAP = uniswap_exchange.get_price_output(token0=DAI,  token1=ETH, qty=1)

print(DAI_SWAP)
```

EDIT: I wonder if it's a gas fee? Can we calculate this/extract this value?

EDIT 2: Looking at another issue I've seen this as a possible solution, could someone explain the exponent work (10** etc) and where these values are obtained from?

```
decimals_usdt = uniswap_exchange.get_token(USDT).decimals
decimals_eth = uniswap_exchange.get_token(ETH).decimals

quantity = 10 ** decimals_eth
price = uniswap_exchange.get_price_input(ETH, USDT, qty=quantity)
price = price / 10 ** decimals_usdt
print(price)
```

Thoughts @ErikBjare?

**Comment by ErikBjare** [comment]
You may not be getting the best route from the pair by default. Try specifying `route` to [`get_price_input`](http://uniswap-python.com/api.html). See https://github.com/uniswap-python/uniswap-python/issues/93 and related discussion for discussion about automating how to find the ~best pair.

> EDIT: I wonder if it's a gas fee? Can we calculate this/extract this value?

You can get the current gas price from the `uniswap_exchange.web3` instance. Check the Web3.py docs.

Note that neither this wrapper nor the Uniswap interface takes gas costs into account when quoted.

> EDIT 2: Looking at another issue I've seen this as a possible solution, could someone explain the exponent work (10** etc) and where these values are obtained from?

The decimals are retrieved from the ERC20 contracts, they are needed to scale the traded amount into their base (non-decimal) unit (`1 ETH = 10**18 WEI`).

**Comment by Marsh-James** [comment]
> You may not be getting the best route from the pair by default. Try specifying `route` to [`get_price_input`](http://uniswap-python.com/api.html). See #93 and related discussion for discussion about automating how to find the ~best pair.
> 
> > EDIT: I wonder if it's a gas fee? Can we calculate this/extract this value?
> 
> You can get the current gas price from the `uniswap_exchange.web3` instance. Check the Web3.py docs.
> 
> Note that neither this wrapper nor the Uniswap interface takes gas costs into account when quoted.
> 
> > EDIT 2: Looking at another issue I've seen this as a possible solution, could someone explain the exponent work (10** etc) and where these values are obtained from?
> 
> The decimals are retrieved from the ERC20 contracts, they are needed to scale the traded amount into their base (non-decimal) unit (`1 ETH = 10**18 WEI`).

Thanks for getting back to me. Comparing with uniswap v3, with multihops disabled, so we're looking at TOKEN->TOKEN directly there's still this discrepancy. Does uniswap do anything off chain perhaps or could the infrua node be lagging in some capacity?

**Comment by liquid-8** [comment]
> could the infrua node be lagging

Yes. Just set up the local node; you can use light-mode for syncing so it requires like 700Mb of the disc space.
